compiler/modules/front/resolveTypes.bal

```diff
@@ -28,6 +28,20 @@ function resolveTypes(ModuleSymbols mod) returns ResolveTypeError? {
     foreach var defn in mod.defns {
         check resolveDefn(mod, defn);
     }
+    if !mod.tc.env.isReady() {
+        // This should never happen
+        panic err:impossible("type environment is not ready");
+    }
+    foreach var { semType, modDefn, startPos, endPos } in mod.deferredEmptinessChecks {
+        check testTypeForFiniteShape(mod.tc, semType, modDefn, startPos, endPos);
+    }
+}
+
+function testTypeForFiniteShape(t:Context tc, t:SemType semType, s:ModuleLevelDefn modDefn, Position startPos, Position endPos) returns ResolveTypeError? {
+    if t:isEmpty(tc, semType) {
+        d:Location loc = s:locationInDefn(modDefn, { startPos, endPos });
+        return err:semantic("invalid recursive type (contains no finite shapes)", loc);
```

Comment:
I think we need to keep track of the original type descriptor and give a different error for the intersection case.

So for the list of deferred types I suggest keeping the type desc instead of the location.

What happens if we have a record type that is empty because it has a field type that is empty? Do we always give the right error in that case.

Reply:
Fixed. We had a problem with cases such as [this](https://github.com/heshanpadmasiri/nballerina/blob/fix-infinite-shapes/compiler/testSuite/14-bug/intersect1-e.bal) since we show only one error message. Depending on whether the record type desc or binary type desc (intersection) got checked first (in the example it used to be, record type T1 will  get checked first since by the time we are resolving that type env will be ready) we will get the error for intersection or infinite type and stop. Fixed by keeping deferred emptiness checks for intersections separate and delaying other checks  until we have gone through all the intersection checks.

Reply:
Currently, I am doing this with a naive approach by

1. In order to find any pending "child" type descriptor I iterate over all the deferred emptiness checks
2. Once the env is ready in `resolveTypes` I sort all the deferred emptiness checks in the descending order of start positions so that inner type desc comes before the outer type desc

The problem with this approach is 1 is slow and 2 doesn't take into account some of the emptiness checks may not be overlapping (as a result in cases like [this](https://github.com/heshanpadmasiri/nballerina/blob/fix-infinite-shapes/compiler/testSuite/14-infinite/infiniteRecord3-e.bal) second type is checked first). I think the correct way to do this is to build a tree/forest structure based on the start and end positions. @jclark do you think it is useful to do that?